scheduler/src/cook/mesos/constraints.clj

```diff
@@ -0,0 +1,307 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.mesos.constraints
+  (:require [clojure.tools.logging :as log]
+            [cook.mesos.util :as util]
+            [cook.mesos.group :as group]
+            [swiss.arrows :refer :all])
+  (import com.netflix.fenzo.VirtualMachineLease))
+
+;; Wisdom:
+;; * This code expects that attributes COOK_GPU? and HOSTNAME are set for all
+;; VMs. This is done by scheduler/get-offer-attr-map , which fetches the mesos
+;; attributes from an offer and also adds some other custom attributes that are
+;; useful to Cook.  In general, adding custom attributes is great for flagging
+;; features of a VM that are hard to figure out during a rebalancing cycle. For
+;; example, figuring out if a machine has gpus requires knowing a VM's
+;; resources, which are not accessible to the rebalancer (would need to cache
+;; resources along with attributes). Instead, we just set the COOK_GPU?
+;; attribute in scheduler/get-offer-attr-map/
+
+(defn get-vm-lease-attr-map
+  "Returns the attribute map of a Fenzo VirtualMachineLease."
+  [^VirtualMachineLease lease]
+  (.getAttributeMap lease))
+
+(defn job-needs-gpus?
+  "Returns true if the provided job needs GPUs, false otherwise."
+  [job]
+  (->> (:job/resource job)
+       (filter (fn gpu-resource? [res]
+                 (and (= (:resource/type res) :resource.type/gpus)
+                      (pos? (:resource/amount res)))))
+       (seq)
+       (boolean)))
+
+;; Job host placement constraints
+(defprotocol JobConstraint
+  "A placement constraint that is defined only by the job being placed."
+  (job-constraint-name [this]
+    "The name of this constraint")
+  (job-constraint-evaluate [this target-vm-resources target-vm-attrs]
+                           [this target-vm-resources target-vm-attrs target-vm-tasks-assigned]
+    "Evaluates whether a vm with resources 'target-vm-resources', attributes 'target-vm-attrs'
+     and TaskRequests (optional) 'target-vm-tasks-assigned' passes the job's constraint. Must return a
+     2-element vector, where the first element is a boolean (whether the constraint was passsed) and
+     the second element is a string (empty string if constraint passed, an explanation otherwise)."))
+
+(defn get-class-name
+  "Returns the name of a class without package information."
+  [object]
+  (last (clojure.string/split (str (type object)) #"\.")))
```

Comment:
Ooh, knew there must have been a better way for this. Now using .getSimpleName :thumbsup: